src/transformStateWithClones.js

```diff
@@ -6,8 +6,32 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
 
+      case 'clear':
+        currentState = {};
+        break;
+    }
+
+    stateHistory.push(currentState);
```

Comment:
You are pushing the reference to the currentState object into the stateHistory array. This will result in all entries in the stateHistory array pointing to the same object, which is the final state. Instead, you should push a copy of currentState to avoid this issue.